### ISSUE TITLE
Code action to create nonexistent files which are linked

### ISSUE THREAD

**Issue by jparoz** [artempyanykh/marksman#219]
It would be great if when making e.g. a wikilink `[[Some file]]`, when `"Some file.md"` doesn't exist, there was a code action to create the file. In my usage, it might be even better if it were possible to configure this to happen automatically when going to definition. This type of behaviour is basically the same as in Obsidian.

Thanks for the great project!

**Comment by artempyanykh** [comment]
@jparoz sure! You pretty much got it!

1. [Server.fs](https://github.com/artempyanykh/marksman/blob/10d38ef4e8dfda36fc91b1bc054c560be94905cd/Marksman/Server.fs#L906-L941) is the entry point into the CodeAction request response.
2. You can implement the actual logic of the code action and put in in `CodeAction.fs`. A sketch is that 
    1. you'd want to lookup a `Cst` element inside the provided range. 
    2. If it's a link^1 you try to resolve it [as it's done in goto-def](https://github.com/artempyanykh/marksman/blob/10d38ef4e8dfda36fc91b1bc054c560be94905cd/Marksman/Server.fs#L802-L805). 
    3. If it can't be resolved then you can generate a code action with a [WorkspaceEdit](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceEdit) of kind `CreateFile`.
3. You probably want to check the [preferred wiki completion style](https://github.com/artempyanykh/marksman/blob/10d38ef4e8dfda36fc91b1bc054c560be94905cd/Marksman/Config.fs#L158) configured for a folder and create a filename for the new file based on it.

^1:  Should be applicable to both types of link – markdown inline and wiki – but you may want to focus on wiki links at the start.

**Comment by jparoz** [comment]
Thanks for the sketch! This is super helpful, and I'm excited to get my hands dirty.

I've been reading through [LanguageServerProtocol/Types.fs](https://github.com/artempyanykh/marksman/blob/10d38ef4e8dfda36fc91b1bc054c560be94905cd/LanguageServerProtocol/Types.fs#L1150) to figure out how to create a file; but it looks to me like that file doesn't match the current version of the spec for [WorkspaceEdit's documentChanges field](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceEdit). Specifically, the `documentChanges` field should allow an array of `(TextDocumentEdit | CreateFile | RenameFile | DeleteFile)`, but it only allows an array of `TextDocumentEdit`.

It seems as though the LanguageServerProtocol code is forked(?) from [another repo](https://github.com/ionide/LanguageServerProtocol/blob/e501c0648562be7e54c26be44f0a723435fa87fe/src/Types.fs#L1248), and the same difference from the spec is present there. Is this something that needs to be addressed there first?

**Comment by artempyanykh** [comment]
> It seems as though the LanguageServerProtocol code is forked(?) from [another repo](https://github.com/ionide/LanguageServerProtocol/blob/e501c0648562be7e54c26be44f0a723435fa87fe/src/Types.fs#L1248)

@jparoz yes, I vendored Ionide.LSP because I needed to make changes often. I try to upstream the changes regularly (as well as pull the upstream changes). I'd recommend just making the necessary changes to LSP Types **inside Marksman**. We can deal with the upstream later.

**Comment by VigneshR387** [comment]
Can this functionality be extended to also specify the directory relative to project root   where the notes should be created? Something like setting  "NewNotes" directory as the default directory to store notes that are created through code action.

**Comment by artempyanykh** [comment]
@VigneshR387 if someone's willing to figure out how this is supposed to work in the context of LSP, then perhaps.
